### PR TITLE
Introduce ConcatRecordReader

### PR DESCRIPTION
**Pull request by LadyForest**
`ConcatRecordReader` is an impl of `RecordReader` to concatenate a list of `RecordReader` together. The key and sequence number already sort records read by the readers, and there's no overlap between the key range for any two different readers. `ConcatRecordReader` is used to concatenate records within a `SortedRun`, or between `Section`s when compaction happens.

### CODE REVIEW

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/merge/ConcatMergeReader.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        }
+        if (queue.size() > 0) {
+            return new ConcatRecordIterator();
+        }
+        return null;
+    }
+
+    @Override
+    public void close() throws IOException {
+        for (RecordReader<KeyValue> reader : sortedReaders) {
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/merge/ConcatMergeReader.java]: I think we should close it when it is empty.

**Review by JingsongLi — COMMENTED** [review]
The purpose of `ConcatReader` is to read data sequentially, it is to avoid opening multiple readers at the same time, because there is no overlap between them.

**Comment by JingsongLi** [comment]
It is definitely not a `MergeReader`.

**Comment by LadyForest** [comment]
> It is definitely not a `MergeReader`.

If we have a strict definition of `merge` that requires consuming data from multiple input readers simultaneously, yes, it is.

I was thinking by `merge` it means the outer reader gathers multiple inner readers together to perform read operation, and how the internal readers are invoked does not matter.

**Comment by JingsongLi** [comment]
> > It is definitely not a `MergeReader`.
> 
> If we have a strict definition of `merge` that requires consuming data from multiple input readers simultaneously, yes, it is.
> 
> I was thinking by `merge` it means the outer reader gathers multiple inner readers together to perform read operation, and how the internal readers are invoked does not matter.

The biggest purpose of `ConcatReader` is to avoid merging, and opening multiple readers will consume a lot of memory Buffer and the file open handles.

**Comment by LadyForest** [comment]
> > > It is definitely not a `MergeReader`.
> > 
> > 
> > If we have a strict definition of `merge` that requires consuming data from multiple input readers simultaneously, yes, it is.
> > I was thinking by `merge` it means the outer reader gathers multiple inner readers together to perform read operation, and how the internal readers are invoked does not matter.
> 
> The biggest purpose of `ConcatReader` is to avoid merging, and opening multiple readers will consume a lot of memory Buffer and the file open handles.

Thanks for the explanation, and I got your point.  The internal readers should be invoked one by one on the implementation side, and we don't define this behavior as a "merge" process. I'll update the code soon.

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/ConcatRecordReader.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    @Nullable
+    @Override
+    public RecordIterator<KeyValue> readBatch() throws IOException {
+        if (sortedReaders.isEmpty()) {
+            return null;
+        }
+        if ((current == null || current.getIndex() < sortedReaders.size()) && nextBatchOrReader()) {
+            return new ConcatRecordIterator();
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/ConcatRecordReader.java]: Here can return `subReader.iterator` directly.

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/ConcatRecordReader.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+import java.util.List;
+
+/**
+ * This reader is to concatenate a list of {@link RecordReader}s and read them sequentially. The
+ * input list is already sorted by key and sequence number, and the key intervals do not overlap
+ * each other.
+ */
+public class ConcatRecordReader implements RecordReader<KeyValue> {
+
+    private final List<RecordReader<KeyValue>> sortedReaders;
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/ConcatRecordReader.java]: I think a `Queue<Supplier<RecordReader<KeyValue>>>` is enough.
`RecordReader` should be create lazily.

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/ConcatRecordReader.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/ConcatRecordReader.java:42]: Another can be a `RecordReader<KeyValue> peekReader` (or topReader) in class member.
Can be simpler version of guava `Iterators.ConcatenatedIterator`.

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/utils/IndexedRecordIterator.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+package org.apache.flink.table.storage.filestore.utils;
+
+import org.apache.flink.table.storage.filestore.KeyValue;
+
+/**
+ * A wrapper iterator of a {@link RecordIterator} with the index in an ordered reader list. The
+ * reader list is already sorted by the key and sequence number. The index helps track which reader
+ * should invoke {@link RecordReader#readBatch()} operation.
+ */
+public class IndexedRecordIterator {
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/utils/IndexedRecordIterator.java]: Revert this?

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/ConcatRecordReader.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    public ConcatRecordReader(List<Supplier<RecordReader<KeyValue>>> sortedReaderSuppliers) {
+        this.queue = new LinkedList<>(sortedReaderSuppliers);
+    }
+
+    @Nullable
+    @Override
+    public RecordIterator<KeyValue> readBatch() throws IOException {
+        Supplier<RecordReader<KeyValue>> supplier;
+        RecordIterator<KeyValue> iterator = null;
+        while (queue.size() > 0
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/ConcatRecordReader.java]: This condition looks magic...
Simple code should:
```
while (true) {
  if (current != null) {
    RecordIterator<KeyValue> iterator = current.readBatch;
    if (iterator != null) {
       return iterator;
    }
    current.close
  }
  current = queue.poll;
  if (current == null) {
     return null;
  }
}
```

Comment (LadyForest) [reply flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/ConcatRecordReader.java]: Hi Jingsong, thanks for the comment. However, I hold the opposite opinion towards the `while ( true ) ` style. 
From my own experience, `while ( true )`  has less readability than `while ( condition )` because I have to pay more effort to check the termination logic that spreads within the body.

And I also searched for some references to support my idea. I quoted some of them here 
> *Quote 1* 
> my experience has been that when I force myself to use the simplest possible control structures (and really think about designing within those constraints), I most often find that the result is simpler, clearer code. The drawback with 
> ```java  
>  while (true) {
>      action;
 >     if (test0) break;
 >     action;
>  }
>  ```
> is that it's easy to let `action0` and `action1` become larger and larger chunks of code, or to add "just one more" test-break-action sequence, until it becomes difficult to point to a specific line and answer the question, "What conditions do I know hold at this point?" So, without making rules for other programmers, I try to avoid the `while (true) {...}` idiom in my own code whenever possible.


> *Quote 2* 
> When you can write your code in the form `while (condition) { ... }` or `while (!condition) { ... }` with **no exits (break, continue, or goto) in the body**, that form is preferred, because someone can read the code and **understand the termination condition just by looking at the header**. That's good.

More discussions can be found at https://stackoverflow.com/questions/390481/is-while-true-with-break-bad-programming-practice

Back to our case, the logic behind the condition is that: 
* firstly the queue should not be empty, and thus we can do a peek
* and the peeked value (the `supplier`) cannot be null and thus we can perform a get operation
* and then if the fetched reader is not null, then we can invoke this reader's `readBatch` operation. 

So what if 
* the supplier is null? 
* or the fetched reader is null? 
* or the iterator returned by the reader's `readBatch` is null? 
It means the current head of the queue is not what we want, we can throw it away, so just pop it. In the meanwhile, what if the read's `readBatch` operation is invoked? We close it by the way. it's pretty clear.

What do you think?

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/utils/IndexedRecordIterator.java (hunk 5 above)
Comment (LadyForest) [reply flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/utils/IndexedRecordIterator.java]: Sure, if we don't have any other circumstances where the index is needed

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/ConcatRecordReader.java (hunk 6 above)
Comment (JingsongLi) [reply flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/ConcatRecordReader.java]: What I strongly disagree with is: the combination of assignment and conditional judgment into one, and the combination of `||` and `&&`.

Comment (JingsongLi) [reply flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/ConcatRecordReader.java]: For this case, it is not a simple condition to continue the loop. This condition is interspersed with many actions.
It's all in the conditions, and I can't even tell if there are any bugs at all.
If a code must need tests to determine whether it has bugs, then I think its readability is problematic.

Comment (JingsongLi) [reply flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/ConcatRecordReader.java]: Why peek in condition?
Why poll? No assignment?
Why `current = supplier.get()` everytime?
Do I still need to go over how it short-circuited?

Comment (JingsongLi) [reply flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/ConcatRecordReader.java]: What do you think about:
```
Iterator iterator = null;
while (iter == null) {
  if (current != null) {
    iterator = current.readBatch;
  } else {
     if (current != null) {
         current.close;
     }
     current = queue.poll;
     if (current == null) {
        return null;
     }
  }
}
```

Comment (LadyForest) [reply flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/ConcatRecordReader.java]: Hi Jingsong, Thanks for your comments. I've made a long reply to you and it may take you several minutes to read it. It contains three parts, if you have a limited time schedule you can skip the first and the third part.

### TL;DR
- I explained the short-circuit logic in part2.
- I think the logic control branch nested in the while loop should be avoided as could as possible.

### Part 1 Reply to the first suggested modification
For the first suggested modification, I've got some questions.
```java
while (true) {
  if (current != null) {
    RecordIterator<KeyValue> iterator = current.readBatch;
    if (iterator != null) {
       return iterator;
    }
    current.close
  }
  current = queue.poll;
  if (current == null) {
     return null;
  }
}

```
* This code snippet will actually end up in an infinite loop because `current` does not get a chance to be assigned within the loop body.
* The type of `queue` is `Supplier<RecordReader<KeyValue>>`, not `RecordReader<KeyValue>` (unless the input parameter is changed to iterator to achieve deferred creation). So `current ` should be `queue.poll().get()`, but the dereference of `queue.poll()` should be checked, the queue may be empty, or the supplier could be null. **Do we need to add a strong assumption that the supplier could not be null?**
* I've made some modifications to make it a worked version.
```java
while (true) {
    if (current != null) {
        RecordIterator<KeyValue> iterator = current.readBatch();
        if (iterator != null) {
            return iterator;
        }
        current.close();
        if (queue.size() > 0) {
            current = queue.poll().get(); // this dereferences on supplier makes assumption that supplier is always not null
        } else {
            // should manually set current to null
            current = null;
        }
        if (current == null) {
            return null;
        }
    } else if (queue.size() > 0) {
            // should assign value to current
            current = queue.poll().get(); // this dereferences on supplier makes assumption that supplier is always not null
    } else {
            return null;
    }
}
```
---

### Part 2 Reply to the questions

> Why peek in condition?

Because I want to make sure that **`supplier` is always a fresh reference pointing to the current head of the queue.** Why? Let's analyze the movement of the head.

There're three conditions that the head should move to the next.
* The supplier is null. **pop, head move**
* The supplier is not null, but the dereferenced reader is null. **pop, head move**
* The supplier is not null and the dereferenced reader is not null, but the iterator of the reader is null. **pop, head move**
Any match of these three conditions will cause a pop operation and make head changes. 

> Why poll? No assignment?
Because entering the loop body means the current head of the queue is useless. Actually, the conditions in the while loop are meant to find a useless head. Once entered this body, the head must be useless (matched the previous three conditions).

> Do I still need to go over how it is short-circuited?

Let's take a more detailed look at the condition.

First step, let's simplify the code to this to make it much clearer.
```java
while (queue.size > 0 && theHeadOfQueueIsUseless()) {
    queue.pop();
}

private boolean theHeadOfQueueIsUseless() {
    // do something 
}
```

This is to illustrate the intention of the while loop => keep finding the useless head and popping it until finding a **qualified head** (the computed iterator is not null) or the queue is empty.

Second step, let's complete the logic in `theHeadOfQueueIsUseless` method.
```java
private boolean theHeadOfQueueIsUseless() {
    Supplier<RecordReader<KeyValue>> supplier = queue.peek(); // we can directly peek because queue.size() > 0 does not be short-curcuited
    if (supplier == null) {
        return true;
    }
    current = supplier.get();
    if (current == null) {
        return true;
    }
    iterator = current.readBatch();
    if (iterator == null) {
        current.close();
        return true;
    }
    return false;
}
```
**The core benefit of this approach (using short-curcuit) is that the head supplier, the dereferenced reader, and the created iterator are always get updated at each iteration round**, and avoid the multiple checks on each other like
```java
// find the first nonnull supplier
while(queue.size() > 0 && supplier == null) {
    supplier = queue.pop();
}
if (supplier != null) {
    current = supplier.get();
    // find first nonnull reader
    while (current == null) {
        // supplier = queue.pop();
        ...
    }
    if (current != null) {
        iterator = current.readBatch();
        if (iterator == null) { .... }
    }
}
```
**And the extra benefit of this approach is that we don't need to care if there are null suppliers when offering the list to the queue.**

It can be simplified to 

```java
private boolean theHeadOfQueueIsUseless() {
    Supplier<RecordReader<KeyValue>> supplier = queue.peek(); 
    return supplier == null || (current = supplier.get()) == null || (iterator = current.readBatch()) == null;
}
```

Does this explanation detailed enough to illustrate the short-circuit logic well? What if I extract a helper method (the method name can be discussed) to help improve the readability? 

---

### Part 3 Reply to the second suggested modification

> What do you think about:
> 
> ```java
> 1    Iterator iterator = null;
> 2    while (iter == null) {
> 3        if (current != null) {
> 4            iterator = current.readBatch;
> 5        } else {
> 6            if (current != null) {
> 7                current.close;
> 8            }
> 9            current = queue.poll;
>10            if (current == null) {
>11                return null;
>12            }
>13        }
>14    }
> ```

Hi Jingsong, thanks for the reply. I have some questions about the quoted code snippet.

* L#3 ~ L#8
    *  When we reach the `else` branch at L#5, how to reach L#6?
   
* L#9 ~ L#12
    *  `Queue` is a type of `Supplier<RecordReader<KeyValue>>`, not `RecordReader<KeyValue>`
    * The impl of `Queue` is `LinkedList` which accepts offering null values, how we can tell `current` is null means `Queue` is empty?

Comment (JingsongLi) [reply flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/ConcatRecordReader.java]: Thanks for the detailed reply, Your in-depth thinking is awesome! I think I can clarify some of my tendencies:.
- `supplier.get()` return null? Just check it (Precondition.checkNotNull?) and throw exception. There should be no cases for that.
- There is no performance requirement (per batch instead of per record), the readability is the first matter.
- `theHeadOfQueueIsUseless` looks not good to me. Because it should be a judgement method, but it assigns a class member `iterator`, it is dangerous.
- My latter pseudo-code can be ignored because there is no code after this loop, it does not need to have condition.

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/ConcatRecordReader.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    @Override
+    public RecordIterator<KeyValue> readBatch() throws IOException {
+        while (true) {
+            if (current != null) {
+                RecordIterator<KeyValue> iterator = current.readBatch();
+                if (iterator != null) {
+                    return iterator;
+                }
+                current.close();
+                current = queue.isEmpty() ? null : queue.poll().get();
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/ConcatRecordReader.java]: just `current = null`?

**Review by JingsongLi — COMMENTED** [review]
Thanks for the update, left last comment

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/ConcatRecordReader.java (hunk 7 above)
Comment (LadyForest) [reply flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/ConcatRecordReader.java]: > just `current = null`?

Why are you so brilliant! Yes!!

**Review by JingsongLi — APPROVED** [review]
+1
